src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,123 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int BIT_SHIFT_VALUE = 1;
+    private static final int ONE_ELEMENT_SIZE = 1;
+    private static final int EMPTY_LIST_VALUE = 0;
+    private static final int FIRST_ELEMENT_INDEX = 0;
+    private static final int NO_SUCH_ELEMENT_INDEX = -1;
+    private static final String NO_SUCH_ELEMENT_MESSAGE = "There is no such element in list.";
+    private static final String INDEX_OUT_OF_BOUNDS_MESSAGE = "This index does not exist in list.";
+
+    private Object[] data;
+    private int size;
+
+    public ArrayList() {
+        data = new Object[DEFAULT_CAPACITY];
+        size = EMPTY_LIST_VALUE;
+    }
+
+    private Object[] grow() {
+        data = data.length > EMPTY_LIST_VALUE
+                ? Arrays.copyOf(data, data.length + (data.length >> BIT_SHIFT_VALUE))
+                : new Object[DEFAULT_CAPACITY];
+        return data;
+    }
+
+    private int getIndexByElement(T element) {
+        for (int i = 0; i < data.length; i++) {
+            if (data[i] == element || (data[i] != null && data[i].equals(element))) {
+                return i;
+            }
+        }
+        return NO_SUCH_ELEMENT_INDEX;
+    }
+
+    private void checkAddByIndex(int index) {
+        if (index < EMPTY_LIST_VALUE || index > size) {
+            throw new ArrayListIndexOutOfBoundsException(INDEX_OUT_OF_BOUNDS_MESSAGE);
+        }
+    }
+
+    private void checkGetByIndex(int index) {
+        if (index < EMPTY_LIST_VALUE || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException(INDEX_OUT_OF_BOUNDS_MESSAGE);
```

Comment:
Here and for other error messages, we can make them more informative, like which indexes, which elements etc.